### ISSUE TITLE
Add toErr, toOk and toSuccess, toFailure to Option

### ISSUE THREAD

**Issue by stephentetley** [flix/flix#1044]
Currently `Validation` and `Result` have functions to convert to `Option`, but `Option` doesn't have functions to convert to `Result` or `Validation`.

`Option` in Scala's standard library has functions `toRight` and `toLeft` to convert `Option` to `Either`.

For `Result` promoting an `Option` to a `Result` without explicit pattern matching would be nice to have for `let*` code.

Bikeshedding...

Are the names `toErr` / `toOk` and `toSuccess` / `toFailure` good enough? Should they include the type as well e.g. `toResultErr` / `toResultOk`?

Alternatively, should they be defined in the `Result` and `Validation` modules as `fromSome` / `fromNone` functions?

**Comment by magnus-madsen** [comment]
> For Result promoting an Option to a Result without explicit pattern matching would be nice to have for let* code.

Yes, I fully agree.

> Are the names toErr / toOk and toSuccess / toFailure good enough? 

Yes, I think so. If they are in the `Option` namespace then it should be clear what `Option.toOk` does.

> Alternatively, should they be defined in the Result and Validation modules as fromSome / fromNone functions?

Good question! I think `toX` functions are most natural. Using an argument from OOP *gasp* it makes sense that if you have an option `o` you may wonder what functions you can apply to it and it seems most "natural" that you would look for such functions in the `Option` namespaces rather than in say `Result` or `Validation`. You could imagine auto-completing on `o.` using UFCS in some future editor, but writing `fromSome` and then getting a bunch of functions like `Result.fromSome` and `Validation.fromSome` would be less intuitive. Ultimately I could see a situation -- when we have type classes- that we can have a single overloaded operation for "transmuting" one value into another, like e.g. `From` and `Into` in Rust. (I have no idea why they have both, but they do say to only implement `From`)

**Comment by stephentetley** [comment]
Thanks Magnus.

I've put them on a new branch along with the lifters.

I've gone with the same arg order as `Option.getWithDefault` - i.e. the Option value first then the default error or "good" value.

e.g:

~~~

///
/// Returns the Option value `Ok(v) if `o` is `Some(v)`. Otherwise returns `Err(e)`.
///
pub def toOk(o: Option[t], e: e): Result[t, e] = match o {
    case None       => Err(e)
    case Some(a)    => Ok(a)
}

///
/// Returns the Option value `Err(e) if `o` is `Some(e)`. Otherwise returns `Ok(d)`.
///
pub def toErr(o: Option[e], d: t): Result[t, e] = match o {
    case None       => Ok(d)
    case Some(e)    => Err(e)
}
    
~~~

**Comment by magnus-madsen** [comment]
> I've gone with the same arg order as Option.getWithDefault - i.e. the Option value first then the default error or "good" value.

Looks good to me.

**Comment by magnus-madsen** [comment]
Fixed by Stephen in #1050 :)
